src/iter.rs

```diff
@@ -92,3 +93,161 @@ fn hex_chars_to_byte(hi: u8, lo: u8) -> Result<u8, HexToBytesError> {
+where
+    I: Iterator<Item = u8>,
+{
+    /// Constructs a new `BytesToHexIter` from a byte iterator.
+    pub fn new(iter: I) -> BytesToHexIter<I> { Self { iter, low: None } }
+}
+
+impl<I> Iterator for BytesToHexIter<I>
+where
+    I: Iterator<Item = u8>,
+{
+    type Item = char;
+
+    fn next(&mut self) -> Option<char> {
+        match self.low {
+            Some(c) => {
+                self.low = None;
+                Some(c)
+            }
+            None => self.iter.next().map(|b| {
+                let (high, low) = byte_to_hex_chars(b);
+                self.low = Some(low);
+                high
+            }),
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let (min, max) = self.iter.size_hint();
+        match self.low {
+            Some(_) => (min * 2 + 1, max.map(|max| max * 2 + 1)),
+            None => (min * 2, max.map(|max| max * 2)),
+        }
+    }
+}
+
+impl<I> DoubleEndedIterator for BytesToHexIter<I>
+where
+    I: DoubleEndedIterator + Iterator<Item = u8>,
+{
+    fn next_back(&mut self) -> Option<char> {
+        match self.low {
+            Some(c) => {
+                self.low = None;
+                Some(c)
+            }
+            None => self.iter.next_back().map(|b| {
+                let (high, low) = byte_to_hex_chars(b);
+                self.low = Some(low);
+                high
+            }),
+        }
+    }
+}
+
+impl<I> ExactSizeIterator for BytesToHexIter<I>
+where
+    I: ExactSizeIterator + Iterator<Item = u8>,
+{
+    fn len(&self) -> usize { self.iter.len() * 2 }
+}
+
+impl<I> FusedIterator for BytesToHexIter<I> where I: FusedIterator + Iterator<Item = u8> {}
+
+/// Returns the (high, low) hex characters encoding `b`.
+fn byte_to_hex_chars(b: u8) -> (char, char) {
+    const HEX_TABLE: [u8; 16] = *b"0123456789abcdef";
+
+    let high = HEX_TABLE[usize::from(b >> 4)];
+    let low = HEX_TABLE[usize::from(b & 0b00001111)];
+
+    (char::from(high), char::from(low))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn encode_byte() {
+        let tcs =
+            vec![(0x00, ('0', '0')), (0x0a, ('0', 'a')), (0xad, ('a', 'd')), (0xff, ('f', 'f'))];
+        for (b, (high, low)) in tcs {
+            assert_eq!(byte_to_hex_chars(b), (high, low));
+        }
+        assert_eq!(byte_to_hex_chars(0x00), ('0', '0'));
+        assert_eq!(byte_to_hex_chars(0x0a), ('0', 'a'));
+        assert_eq!(byte_to_hex_chars(0xad), ('a', 'd'));
+        assert_eq!(byte_to_hex_chars(0xff), ('f', 'f'));
+    }
+
+    #[test]
+    fn decode_iter_forward() {
+        let hex = "deadbeef";
+        let v = vec![0xde, 0xad, 0xbe, 0xef];
+
+        for (i, b) in HexToBytesIter::new(hex).unwrap().enumerate() {
+            assert_eq!(b.unwrap(), v[i]);
+        }
+    }
+
+    #[test]
+    fn decode_iter_backward() {
+        let hex = "deadbeef";
+        // TODO: Confirm that this is what we want, the digits are still ordered (high, low)?
```

Comment:
In 33e81052babc5712adfe671b2df01a239e936aab:

Yes, this is what we want. Thank you for adding a test.